src/tests/mlkem.rs

```diff
@@ -0,0 +1,257 @@
+         C03E832EFD8AB15C4C5F16CB84BA848A72E09D88777F3B969972D21FD4F60E12\
+         71579D32AB9A1012C9DB0D3204AAF0347925AC89B256B6AA5073CDE02584E602\
+         6ADED3696366D43E5543362749864CDB22E69A18B0124A609BE9D1A0F93C3603\
+         CD5EAB26522E637FD086CE652D0ED7EA525DBC304FC9A5ED64809AAC797964D5\
+         C15FFEF9931AAEE0388863CFBD3310B501930CC545C9891CF103509F88A9300E",
+    )
+    .expect("failed to decode ML-KEM private key");
+    let priv_handle = ret_or_panic!(import_object(
+        session,
+        CKO_PRIVATE_KEY,
+        &[
+            (CKA_KEY_TYPE, CKK_ML_KEM),
+            (CKA_PARAMETER_SET, CKP_ML_KEM_512),
+        ],
+        &[
+            (CKA_VALUE, &pk),
+            (CKA_LABEL, format!("ML-KEM private key").as_bytes()),
+        ],
+        &[(CKA_DECAPSULATE, true)],
+    ));
+
+    let key_template = make_attr_template(
+        &[
+            (CKA_CLASS, CKO_SECRET_KEY),
+            (CKA_KEY_TYPE, CKK_AES),
+            (CKA_VALUE_LEN, 16),
+        ],
+        &[],
+        &[
+            (CKA_ENCRYPT, true),
+            (CKA_DECRYPT, true),
+            (CKA_EXTRACTABLE, true),
+        ],
+    );
+
+    let mut mechanism: CK_MECHANISM = CK_MECHANISM {
+        mechanism: CKM_ML_KEM,
+        pParameter: std::ptr::null_mut(),
+        ulParameterLen: 0,
+    };
+    let mut ciphertext = [0u8; 800];
+    let mut outlen: CK_ULONG = 800;
+    let mut handle_enc = CK_INVALID_HANDLE;
+    let ret = fn_encapsulate_key(
+        session,
+        &mut mechanism,
+        pub_handle,
+        key_template.as_ptr() as *mut _,
+        key_template.len() as CK_ULONG,
+        ciphertext.as_mut_ptr(),
+        &mut outlen,
+        &mut handle_enc,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    let mut handle_dec = CK_INVALID_HANDLE;
+    let ret = fn_decapsulate_key(
+        session,
+        &mut mechanism,
+        priv_handle,
+        key_template.as_ptr() as *mut _,
+        key_template.len() as CK_ULONG,
+        ciphertext.as_mut_ptr(),
+        outlen,
+        &mut handle_dec,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    let mut value_enc = vec![0u8; 16];
+    let mut extract_template = make_ptrs_template(&[(
+        CKA_VALUE,
+        void_ptr!(value_enc.as_mut_ptr()),
+        value_enc.len(),
+    )]);
+
+    let ret = fn_get_attribute_value(
+        session,
+        handle_enc,
+        extract_template.as_mut_ptr(),
+        extract_template.len() as CK_ULONG,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    let mut value_dec = vec![0u8; 16];
+    let mut extract_template = make_ptrs_template(&[(
+        CKA_VALUE,
+        void_ptr!(value_dec.as_mut_ptr()),
+        value_dec.len(),
+    )]);
+
+    let ret = fn_get_attribute_value(
+        session,
+        handle_dec,
+        extract_template.as_mut_ptr(),
+        extract_template.len() as CK_ULONG,
+    );
+    assert_eq!(ret, CKR_OK);
+
+    assert_eq!(value_enc, value_dec);
+
+    /* try again with key generation first */
```

Comment:
I would just throw this into separate test so it can run in parallel. And would give it a try with some different PARAMETER_SET than above just to increase the coverage.

Reply:
I am adding test vectors from NIST that should try all parameter sets, so I will keep this test linear as is.